server/app/services/grid_service_instance_deployer.rb

```diff
@@ -4,82 +4,117 @@ class GridServiceInstanceDeployer
-    wait_for_service_state(service_instance, 'running', deploy_rev)
-
-    true
-  rescue => exc
-    error "failed to deploy service instance #{self.grid_service.to_path}-#{instance_number} to node #{node.name}"
-    error exc.message
-    error exc.backtrace.join("\n")
-    false
+    ensure_service_instance(service_instance, @host_node, deploy_rev, 'running')
+
+  rescue => error
+    warn "Failed to deploy service instance #{@grid_service.to_path}-#{@instance_number} to node #{@host_node.name}: #{error.class}: #{error}\n#{error.backtrace.join("\n")}"
+    @grid_service_instance_deploy.set(:deploy_state => :error, :error => "#{error.class}: #{error}")
+  else
+    @grid_service_instance_deploy.set(:deploy_state => :success)
   end
 
-  # @param [GridServiceInstance] current_instance
-  def stop_current_instance(current_instance)
-    current_instance.set(desired_state: 'stopped')
-    if current_instance.host_node && current_instance.host_node.connected?
-      notify_node(current_instance.host_node)
-      wait_for_service_state(current_instance, 'stopped')
-    end
+  # @param service_instance [GridServiceInstance]
+  # @param deploy_rev [String]
+  def stop_current_instance(service_instance, deploy_rev)
+    info "Stopping existing service service #{@grid_service.to_path} instance #{service_instance.instance_number} on previous node #{service_instance.host_node.name}..."
+
+    ensure_service_instance(service_instance, service_instance.host_node, deploy_rev, 'stopped')
+
+  rescue => error
+    warn "Failed to stop existing service #{@grid_service.to_path} instance #{service_instance.instance_number} on previous node #{service_instance.host_node.name}: #{error}"
   end
 
-  # @param [GridServiceInstance] service_instance
-  def wait_for_service_state(service_instance, state, rev = nil)
-    wait_until!("service #{@grid_service.to_path} instance #{service_instance.instance_number} is #{state} on node #{service_instance.host_node.to_path}", timeout: 300) do
+  # Update service instance, notify node, wait for update, and ensure that service is in desired state.
+  # Returns updated GridServiceInstance, or raises on any errors.
+  #
+  # @param service_instance [GridServiceInstance]
+  # @param node [String] move to host node
+  # @param deploy_rev [String] revision
+  # @param desired_state [String] set and check for container state
+  # @raise [RpcClient::TimeoutError] notify node failed
+  # @raise [AgentError] agent failed to apply update
+  # @raise [StateError] unexpected state after update
+  # @raise [Timeout::Error] no update from agent
+  # @return [GridServiceInstance]
+  def ensure_service_instance(service_instance, node, deploy_rev, desired_state)
+    raise NodeError, "Host node is missing" unless node
+    raise NodeError, "Host node is offline" unless node.connected?
+
+    service_instance.set(
+      host_node_id: node.id,
+      deploy_rev: deploy_rev,
+      desired_state: desired_state,
+      rev: nil, # reset
```

Comment:
This is problematic because both the `ensure_service_instance(..., 'stopped')` and `ensure_service_instance(..., 'running')` happen with the same `deploy_rev`... Ideally, each `ensure_service_instance` operation would just set its own `deploy_rev: Time.now.utc` (XXX: with sub-second percision?), but then I'm not sure what affect that would have on the actual Docker container `deploy_rev`?